core/vm/ewasm.go

```diff
@@ -77,60 +78,69 @@ func (in *InterpreterEWASM) Run(contract *Contract, input []byte) ([]byte, error
 	}
+
+	// The module should not have any start function
+	if module.Start != nil {
+		return nil, fmt.Errorf("A contract should not have a start function: found #%d", module.Start.Index)
+	}
+
 	vm, err := exec.NewVM(module)
 	if err != nil {
 		return nil, fmt.Errorf("could not create the vm: %v", err)
 	}
 	vm.RecoverPanic = true
 	in.vm = vm
 
-	// If the module doesn't have any start function, check if the
-	// exports specify a `main` function and if so, execute it.
-	if module.Start == nil {
-		var mainIndex = -1
-		for name, entry := range module.Export.Entries {
-			if name == "main" && entry.Kind == wasm.ExternalFunction {
-
-				// Check input and output types
-				sig := module.FunctionIndexSpace[entry.Index].Sig
-				if len(sig.ParamTypes) == 0 && len(sig.ReturnTypes) == 0 {
-					mainIndex = int(entry.Index)
-					break
+	// Look for the "main" function and execute it after checking it
+	// has the right kind of signature.
+	for name, entry := range module.Export.Entries {
+		if name == "main" && entry.Kind == wasm.ExternalFunction {
+
+			// Check input and output types
+			sig := module.FunctionIndexSpace[entry.Index].Sig
+			if len(sig.ParamTypes) == 0 && len(sig.ReturnTypes) == 0 {
+				_, err = vm.ExecCode(int64(entry.Index))
+
+				if in.StateDB.HasSuicided(contract.Address()) {
+					if initialGas-contract.Gas-params.TxGas < 2*params.SuicideRefundGas {
+						in.StateDB.AddRefund((initialGas - contract.Gas - params.TxGas) / 2)
+					} else {
+						in.StateDB.AddRefund(params.SuicideRefundGas)
+					}
+					err = nil
 				}
-			}
-		}
-		if mainIndex >= 0 {
-			_, err = vm.ExecCode(int64(mainIndex))
-
-			if in.StateDB.HasSuicided(contract.Address()) {
-				if initialGas-contract.Gas-params.TxGas < 2*params.SuicideRefundGas {
-					in.StateDB.AddRefund((initialGas - contract.Gas - params.TxGas) / 2)
-				} else {
-					in.StateDB.AddRefund(params.SuicideRefundGas)
+
+				if err != nil {
+					in.terminationType = TerminateInvalid
 				}
-				err = nil
-			}
 
-			if err != nil {
-				in.terminationType = TerminateInvalid
+				return in.returnData, err
 			}
-		} else {
-			return nil, errors.New("no default entry available in the module, and none were specified")
-		}
 
+			// Found a main but it doesn't have the right signature - fail
+			break
+		}
 	}
 
-	return in.returnData, err
+	return nil, errors.New("Could not find a suitable 'main' function in that contract")
 }
 
-// CanRun checks the binary for a WASM header. ReadModule should fail fairly
-// quickly if it doesn't fine the \0asm header, so in theory calling is should
-// be fairly cheap.
+// CanRun checks the binary for a WASM header and accepts the binary blob
+// if it matches.
 func (in *InterpreterEWASM) CanRun(file []byte) bool {
-	// TODO save the module to a list, to avoid a double-load.
-	_, err := wasm.ReadModule(bytes.NewReader(file), WrappedModuleResolver(in))
+	// Check the header
+	if len(file) <= 8 || string(file[:4]) != "\000asm" {
```

Comment:
Please also include the wasm version, I think. (Need to specify this fully.)

Hera includes the version: https://github.com/ewasm/hera/blob/master/src/helpers.cpp#L89-L100